src/main/java/uk/gov/justice/hmpps/prison/repository/jpa/model/OffenderBooking.java

```diff
@@ -177,6 +193,167 @@
+                .earlyTermDate(sc.getEarlyTermDate())
+                .midTermDate(sc.getMidTermDate())
+                .lateTermDate(sc.getLateTermDate())
+                .topupSupervisionExpiryDate(sc.getTopupSupervisionExpiryDate())
+                .tariffDate(sc.getTariffDate())
+                .dtoPostRecallReleaseDate(sc.getDprrdCalculatedDate())
+                .dtoPostRecallReleaseDateOverride(sc.getDprrdOverridedDate())
+                .nonParoleDate(sc.getNpdCalculatedDate())
+                .nonParoleOverrideDate(sc.getNpdOverridedDate())
+                .nonDtoReleaseDate(sc.getNonDtoReleaseDate())
+                .nonDtoReleaseDateType(sc.getNonDtoReleaseDateType())
+                .releaseDate(getReleaseDate())
+                .confirmedReleaseDate(getConfirmedReleaseDate().orElse(null))
+                .build())
+            .orElse(
+                SentenceCalcDates.sentenceCalcDatesBuilder()
+                    .bookingId(getBookingId())
+                    .sentenceStartDate(getSentenceStartDate().orElse(null))
+                    .additionalDaysAwarded(getAdditionalDaysAwarded())
+                    .releaseDate(getReleaseDate())
+                    .confirmedReleaseDate(getConfirmedReleaseDate().orElse(null))
+                    .build());
+    }
+
+    public record DerivedKeyDates(NonDtoReleaseDate nonDtoReleaseDate, LocalDate releaseDate) {
+    }
+
+    public static DerivedKeyDates deriveKeyDates(final KeyDateValues keyDateValues) {
+
+        // Determine non-DTO release date
+        final var nonDtoReleaseDate = SentenceCalculation.deriveNonDtoReleaseDate(keyDateValues).orElse(null);
+
+        // Determine offender release date
+        final var releaseDate = deriveOffenderReleaseDate(keyDateValues, nonDtoReleaseDate);
+
+        return new DerivedKeyDates(nonDtoReleaseDate, releaseDate);
+    }
+
+    /**
+     * Offender release date is determined according to algorithm.
+     * <p>
+     * 1. If there is a confirmed release date, the offender release date is the confirmed release date.
+     * <p>
+     * 2. If there is no confirmed release date for the offender, the offender release date is either the actual
+     * parole date or the home detention curfew actual date.
+     * <p>
+     * 3. If there is no confirmed release date, actual parole date or home detention curfew actual date for the
+     * offender, the release date is the later of the nonDtoReleaseDate or midTermDate value (if either or both
+     * are present).
+     *
+     * @param keyDateValues     a set of key date values used to determine the Non deterministic release date
+     * @param nonDtoReleaseDate derived Non deterministic release date information
+     * @return releaseDate
+     */
+    private static LocalDate deriveOffenderReleaseDate(final KeyDateValues keyDateValues, final NonDtoReleaseDate nonDtoReleaseDate) {
+
+        final LocalDate releaseDate;
```

Comment:
Feels like thee should be unit for this

Reply:
Release Dates is only tested as integration cucumber tests at the moment